tracer/build/_build/CodeGenerators/EquatableArray.cs

```diff
@@ -0,0 +1,121 @@
+// <copyright file="EquatableArray.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace CodeGenerators;
+
+/// <summary>
+/// An immutable, equatable array. This is equivalent to <see cref="Array"/> but with value equality support.
+/// </summary>
+/// <typeparam name="T">The type of values in the array.</typeparam>
+public readonly struct EquatableArray<T> : IEquatable<EquatableArray<T>>, IEnumerable<T>
+    where T : IEquatable<T>
+{
+    /// <summary>
+    /// The underlying <typeparamref name="T"/> array.
+    /// </summary>
+    private readonly T[]? _array;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="EquatableArray{T}"/> struct.
+    /// </summary>
+    /// <param name="array">The input array to wrap.</param>
+    public EquatableArray(T[] array)
+    {
+        _array = array;
+    }
+
+    /// <summary>
+    /// Gets the length of the array, or 0 if the array is null
+    /// </summary>
+    public int Count => _array?.Length ?? 0;
+
+    /// <summary>
+    /// Checks whether two <see cref="EquatableArray{T}"/> values are the same.
+    /// </summary>
+    /// <param name="left">The first <see cref="EquatableArray{T}"/> value.</param>
+    /// <param name="right">The second <see cref="EquatableArray{T}"/> value.</param>
+    /// <returns>Whether <paramref name="left"/> and <paramref name="right"/> are equal.</returns>
+    public static bool operator ==(EquatableArray<T> left, EquatableArray<T> right)
+    {
+        return left.Equals(right);
+    }
+
+    /// <summary>
+    /// Checks whether two <see cref="EquatableArray{T}"/> values are not the same.
+    /// </summary>
+    /// <param name="left">The first <see cref="EquatableArray{T}"/> value.</param>
+    /// <param name="right">The second <see cref="EquatableArray{T}"/> value.</param>
+    /// <returns>Whether <paramref name="left"/> and <paramref name="right"/> are not equal.</returns>
+    public static bool operator !=(EquatableArray<T> left, EquatableArray<T> right)
+    {
+        return !left.Equals(right);
+    }
+
+    /// <inheritdoc/>
+    public bool Equals(EquatableArray<T> array)
+    {
+        return AsSpan().SequenceEqual(array.AsSpan());
+    }
+
+    /// <inheritdoc/>
+    public override bool Equals(object? obj)
+    {
+        return obj is EquatableArray<T> array && Equals(this, array);
```

Comment:
Does `Equals(this, array)` _definitely_ call the right overload and avoid the infinite recurrsion? I know this one definitely does:
```suggestion
        return obj is EquatableArray<T> array && this.Equals(array);
```

Reply:
Ok. Accepted. Although I copied the implementation in the source generator.

Reply:
Yeah, I think the source generator is wrong, because I wrote that before I knew there was a bug 😅